### PR TITLE
Implement weaver-plugins crate with tests and integration

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Adds the new weaver-plugins crate, including core types for plugin error handling, IPC protocol, manifest, registry, process execution, and a plugin runner with test doubles. 
- Wire-up and integration work to expose plugin orchestration to the weaverd daemon, including workspace wiring and routing support for refactor operations.
- Introduces unit tests for all modules and a BDD feature with step definitions to exercise plugin execution paths.
- Updates docs and exec plan to reflect the new crate and design decisions.

## Changes
- New crate: weaver-plugins
  - error module (PluginError) with descriptive variants and test coverage.
  - protocol module (IPC types): PluginRequest, PluginResponse, FilePayload, PluginOutput, PluginDiagnostic, DiagnosticSeverity.
  - manifest module: PluginKind and PluginManifest with validation and accessors.
  - registry module: PluginRegistry with registration, lookup, and language/kind filters.
  - process module: SandboxExecutor implementing plugin execution via weaver-sandbox.
  - runner module: PluginRunner with a PluginExecutor trait to enable test doubles.
  - lib.rs: public API surface re-exports (PluginError, PluginManifest, protocol types, PluginRegistry, PluginRunner).
  - tests: module-level tests scaffold for core behaviour.
- Routing and integration
  - weaverd workspace updated to include weaver-plugins as a member.
  - Wire-up scaffolding for weaver-plugins in weaverd (reference to PluginRegistry, PluginRunner) and router integration for refactor pathway (act refactor).
  - docs/execplans/3-1-1-weaver-plugins-crate.md added to describe the design decisions and plan.
- Tests
  - crates/weaver-plugins/src/error/tests.rs validating message contents and Send+Sync properties for Io errors.
  - crates/weaver-plugins/src/manifest/tests.rs and protocol/tests.rs cover construction, validation, serde round-trips, and edge cases.
  - crates/weaver-plugins/src/registry/tests.rs and crates/weaver-plugins/src/runner/tests.rs for registry logic and executor delegation.
  - crates/weaver-plugins/src/tests/mod.rs and crates/weaver-plugins/src/tests/behaviour.rs provide crate-level integration tests and BDD-style behaviour tests.
  - crates/weaver-plugins/tests/features/plugin_execution.feature added with scenarios around successful execution, not-found, non-zero exit, empty output, and language-based queries.
- Documentation
  - docs/users-guide.md updated to include act refactor command usage and plugin system overview.
  - docs/weaver-design.md updated with Phase 3.1.1 decisions and IPC protocol rationale.
  - docs/roadmap.md updated to reflect completed entry for plugin crate design.

## Design decisions
- IPC protocol: JSONL over stdio (one-shot per interaction). A single line request is sent to the plugin's stdin, and a single line response is read from stdout. This keeps the model simple for short-lived, sandboxed plugins.
- File content in-band: FilePayload carries full content to avoid filesystem access inside sandboxed plugins, simplifying sandbox permissions and security boundary.
- Executor abstraction: PluginExecutor trait enables production (SandboxExecutor) and test doubles for unit/BDD tests.
- Safety harness alignment: Actuator plugin outputs will flow through the existing Double-Lock safety harness path (when wired in Phase 3.2).

## Test plan
- Unit tests for all modules in weaver-plugins (error, protocol, manifest, registry, process, runner).
- BDD tests using rstest-bdd v0.5.0 (scenarios wired in crates/weaver-plugins/tests/features/plugin_execution.feature).
- End-to-end capability: integration with weaverd (act refactor path) prepared; runtime wiring to execute plugins via PluginRunner will be completed in Phase 3.2.
- Ensure workspace tests pass: cargo test -p weaver-plugins and cargo test across the workspace.

## How to test locally
- Build and run tests for the new crate:
  - cargo test -p weaver-plugins
- Run full workspace tests (may take longer):
  - cargo test --workspace
- Review docs and exec plan:
  - docs/weaver-design.md
  - docs/users-guide.md
  - docs/execplans/3-1-1-weaver-plugins-crate.md

## Migration/compatibility notes
- This PR introduces a new crate and wiring rather than changing existing public APIs. It adds new workspace dependencies and opts into rstest-bdd v0.5.0 for BDD tests. If upgrading BDD fixtures in existing tests, follow the staged plan in docs/execplans and migrate tests accordingly.

## Documentation and follow-ups
- Complete wiring the act refactor handler in weaverd to use PluginRunner.execute() and propagate PluginOutput (Diff) into ContentTransaction as per the final safety harness integration plan.
- Expand docs/users-guide with concrete examples of plugin manifests, usage, and troubleshooting.
- Implement full integration tests that exercise real sandboxed plugin execution in a controlled test environment.

## Acceptance criteria
- [x] New weaver-plugins crate compiles with all modules.
- [x] Unit tests compile and pass for all new modules.
- [x] BDD feature and steps compile and integrate with rstest-bdd v0.5.0.
- [x] Weaverd integration scaffolding updated; router wired for refactor path.
- [x] Documentation updated to reflect design decisions and plan.
- [x] End-to-end test plan documented and prepared for Stage C/D work.

◳ Generated by [DevBoxer](https://www.devboxer.com) ◰

---

ℹ️ Tag @devboxerhub to ask questions and address PR feedback

📎 **Task**: https://www.devboxer.com/task/5a5853af-1c35-466a-99bd-a7e70661ec4f

## Summary by Sourcery

Introduce the weaver-plugins crate for plugin orchestration, wire initial refactor routing into weaverd, and document the plugin system and execution plan.

New Features:
- Add the weaver-plugins crate providing plugin manifests, IPC protocol types, a registry, a sandbox-based executor, and a plugin runner abstraction for plugin orchestration.
- Introduce the act refactor command in weaverd, including argument parsing, workspace file resolution, and scaffolding of PluginRequest construction without executing plugins yet.

Enhancements:
- Integrate weaver-plugins into the workspace and weaverd dependencies for future plugin execution wiring.
- Extend the users guide and design docs with plugin system architecture, IPC protocol details, and refactor command usage.
- Mark the roadmap entry for designing and implementing the weaver-plugins crate as complete and add an execution-plan document describing Phase 3.1.1.

Build:
- Add the weaver-plugins crate to the Cargo workspace members and enable it as a dependency of weaverd.
- Upgrade rstest-bdd and rstest-bdd-macros to version 0.5.0 across the workspace.

Tests:
- Add unit tests for plugin error handling, manifest validation, registry queries, protocol serialization, runner orchestration, and a crate-level end-to-end stub test.
- Add BDD-style tests and a Gherkin feature covering plugin execution behaviour via mock executors, including success, error, and lookup scenarios.
- Extend router tests to cover invalid act refactor argument handling.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Excess Number of Function Arguments
crates/weaver-plugins/src/manifest/mod.rs: PluginManifest.new

### What lead to degradation?
PluginManifest.new has 5 arguments, max arguments = 4

### Why does this problem occur?
Functions with many arguments indicate either a) low cohesion where the function has too many responsibilities, or b) a missing abstraction that encapsulates those arguments.
The threshold for the Rust language is 4 function arguments.

### How to fix it?
Start by investigating the responsibilities of the function. Make sure it doesn't do too many things, in which case it should be split into smaller and more cohesive functions. Consider the refactoring [INTRODUCE PARAMETER OBJECT](https://refactoring.com/catalog/introduceParameterObject.html) to encapsulate arguments that refer to the same logical concept.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Complex Method
crates/weaverd/src/dispatch/act/refactor/mod.rs: parse_refactor_args

### What lead to degradation?
parse_refactor_args has a cyclomatic complexity of 9, threshold = 9

### Why does this problem occur?
A Complex Method has a high cyclomatic complexity. The recommended threshold for the Rust language is a cyclomatic complexity lower than 9.

### How to fix it?
There are many reasons for Complex Method. Sometimes, another design approach is beneficial such as a) modeling state using an explicit state machine rather than conditionals, or b) using table lookup rather than long chains of logic. In other scenarios, the function can be split using [EXTRACT FUNCTION](https://refactoring.com/catalog/extractFunction.html). Just make sure you extract natural and cohesive functions. Complex Methods can also be addressed by identifying complex conditional expressions and then using the [DECOMPOSE CONDITIONAL](https://refactoring.com/catalog/decomposeConditional.html) refactoring.

### Helpful refactoring examples
To get a general understanding of what this code health issue looks like - and how it might be addressed - we have prepared some diffs for illustrative purposes.

#### SAMPLE

~~~diff
# complex_method.js
 function postItem(item) {
   if (!item.id) {
-    if (item.x != null && item.y != null) {
-      post(item);
-    } else {
-      throw Error("Item must have x and y");
-    }
+    // extract a separate function for creating new item
+    postNew(item);
   } else {
-    if (item.x < 10 && item.y > 25) {
-      put(item);
-    } else {
-      throw Error("Item must have an x and y value between 10 and 25");
-    }
+    // and one for updating existing items
+    updateItem(item);
   }
 }
+
+function postNew(item) {
+  validateNew(item);
+  post(item);
+}
+
+function updateItem(item) {
+  validateUpdate(item);
+  put(item);
+}
+
~~~

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Code Duplication
crates/weaver-plugins/src/tests/behaviour.rs: The module contains 4 functions with similar structure: given_actuator,given_sensor,then_output_is_empty,then_success
crates/weaver-plugins/src/manifest/tests.rs: The module contains 2 functions with similar structure: validate_rejects_empty_name,validate_rejects_relative_executable
crates/weaver-plugins/src/protocol/tests.rs: The module contains 2 functions with similar structure: failure_response_round_trip,success_response_round_trip
crates/weaver-plugins/src/registry/mod.rs: The module contains 2 functions with similar structure: PluginRegistry.find_actuator_for_language,PluginRegistry.find_for_language
crates/weaver-plugins/src/tests/behaviour.rs: The module contains 4 functions with similar structure: given_actuator,given_sensor,then_output_is_empty,then_success

### What lead to degradation?


### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Code Duplication
crates/weaver-plugins/src/error/tests.rs: 

### What lead to degradation?
The module contains 2 functions with similar structure: non_zero_exit_error_includes_status,timeout_error_message_includes_duration

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Code Duplication
crates/weaver-plugins/src/tests/behaviour.rs: 

### What lead to degradation?
The module contains 4 functions with similar structure: given_actuator,given_sensor,then_output_is_empty,then_success

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

File: crates/weaver-plugins/src/process.rs [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+            message: String::from("failed to capture stdout"),
+            source: None,
+        })?;
+
+    let stderr = child.stderr.take();
+
+    write_request(name, stdin, request)?;
+    let response_line = read_response(name, stdout, manifest.timeout_secs())?;
+    drain_stderr(name, stderr);
+    wait_for_exit(name, &mut child, manifest.timeout_secs())?;
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-plugins/src/process.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/weaverd/src/dispatch/act/refactor/mod.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+fn resolve_file(workspace_root: &Path, file: &str) -> Result<std::path::PathBuf, DispatchError> {
+    let path = std::path::Path::new(file);
+    if path.is_absolute() {
+        return Err(DispatchError::invalid_arguments(
+            "absolute file paths are not allowed; use a path relative to the workspace root",
+        ));
+    }
+    let resolved = workspace_root.join(path);
+    if !resolved.starts_with(workspace_root) {
+        return Err(DispatchError::invalid_arguments(
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaverd/src/dispatch/act/refactor/mod.rs:198]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- In `process.rs`, `read_response` (and `drain_stderr`) perform blocking reads without a reliable timeout mechanism, while `wait_for_exit` separately enforces a timeout; consider consolidating timeout handling so that all blocking I/O (stdout/stderr reads as well as process exit) respects the same deadline and cannot hang indefinitely if a plugin never writes or closes its pipes.
- The BDD test world in `src/tests/behaviour.rs` still uses `RefCell<TestWorld>` even though rstest-bdd 0.5 supports mutable world fixtures; you can simplify the step signatures to take `&mut TestWorld` directly and drop the `RefCell` indirection.

## Individual Comments

### Comment 1
<location> `crates/weaver-plugins/src/process.rs:194` </location>
<code_context>
+}
+
+/// Drains stderr to avoid blocking the child on a full pipe buffer.
+fn drain_stderr(name: &str, stderr_handle: Option<impl Read>) {
+    let Some(reader) = stderr_handle else {
+        return;
</code_context>

<issue_to_address>
**issue (bug_risk):** The `Option<impl Read>` parameter is not valid Rust syntax and will not compile.

`impl Trait` can’t be used directly inside another generic type (like `Option<impl Read>`) in stable Rust, so this won’t compile. Consider either:

- Making the function generic: `fn drain_stderr<R: Read>(name: &str, stderr_handle: Option<R>)`, or
- Using a trait object: `fn drain_stderr(name: &str, stderr_handle: Option<Box<dyn Read>>)`.

For draining the child’s stderr, the generic form is likely simplest and avoids boxing.
</issue_to_address>

### Comment 2
<location> `crates/weaver-plugins/src/process.rs:183-184` </location>
<code_context>
+        });
+    }
+
+    if elapsed > timeout {
+        return Err(PluginError::Timeout {
+            name: name.to_owned(),
+            timeout_secs,
</code_context>

<issue_to_address>
**issue (bug_risk):** `read_response` can return a timeout error without terminating the still-running child process.

When `read_response` returns `PluginError::Timeout`, `execute_in_sandbox` exits before `wait_for_exit` runs, leaving the child process alive and potentially leaking processes if a plugin is slow or hangs.

Please either:
- Centralize timeout handling in `wait_for_exit` and remove the timeout check from `read_response`, or
- Ensure a timeout from `read_response` also kills and reaps the child (e.g., by reusing the existing `child.kill()` timeout path).

Also, the comment above `read_response` says the timeout is handled by `wait_for_exit`, but the logic is currently split and only `wait_for_exit` performs cleanup.
</issue_to_address>

### Comment 3
<location> `crates/weaverd/src/dispatch/act/refactor/mod.rs:159-168` </location>
<code_context>
+}
+
+/// Resolves a file path relative to the workspace root.
+fn resolve_file(workspace_root: &Path, file: &str) -> Result<std::path::PathBuf, DispatchError> {
+    let path = std::path::Path::new(file);
+    if path.is_absolute() {
+        return Err(DispatchError::invalid_arguments(
+            "absolute file paths are not allowed; use a path relative to the workspace root",
+        ));
+    }
+    let resolved = workspace_root.join(path);
+    if !resolved.starts_with(workspace_root) {
+        return Err(DispatchError::invalid_arguments(
+            "path traversal is not allowed",
+        ));
+    }
+    Ok(resolved)
+}
</code_context>

<issue_to_address>
**🚨 issue (security):** The path traversal check does not prevent `..`-based escapes from the workspace.

The current `starts_with` check is bypassable because it’s purely prefix-based and doesn’t resolve `..` segments, so inputs like `../../etc/passwd` can still pass while escaping the workspace.

To actually confine access to `workspace_root`, resolve the joined path first (e.g. with `canonicalize` or equivalent normalization) and then compare the normalized path against `workspace_root` before accepting it. As written, a malicious relative path can reach arbitrary files outside the workspace.
</issue_to_address>

### Comment 4
<location> `crates/weaver-plugins/src/protocol/tests.rs:14-23` </location>
<code_context>
+
+use super::*;
+
+#[test]
+fn not_found_error_message_includes_name() {
+    let error = PluginError::NotFound {
</code_context>

<issue_to_address>
**suggestion (testing):** Add a test to ensure `diagnostics` defaults correctly when omitted in JSON

Current round-trip tests only cover serialization from the Rust type, which always includes `diagnostics`. Since the struct uses `#[serde(default)]` on this field, please add a test that deserializes minimal JSON without the `diagnostics` key (for both success and failure) and asserts that `diagnostics()` returns an empty slice. This will catch regressions in the serde attributes and preserve compatibility with plugins that omit this field.
</issue_to_address>

### Comment 5
<location> `crates/weaver-plugins/src/error/tests.rs:10-19` </location>
<code_context>
+    );
+}
+
+#[test]
+fn executable_not_found_includes_path() {
+    let error = PluginError::ExecutableNotFound {
+        name: "missing".into(),
+        path: PathBuf::from("/usr/bin/missing-plugin"),
+    };
+    let message = error.to_string();
+    assert!(
+        message.contains("/usr/bin/missing-plugin"),
+        "expected path in message: {message}"
+    );
+}
+
+#[test]
+fn manifest_error_message_is_passthrough() {
+    let error = PluginError::Manifest {
+        message: "name must not be empty".into(),
</code_context>

<issue_to_address>
**suggestion (testing):** Consider adding tests for `DeserializeResponse` and `InvalidOutput` variants to pin their messages

These variants aren’t currently exercised. Since their `Display` output is user-facing when plugins misbehave, it’d be valuable to add focused tests that construct each variant and assert on the critical parts of the message (plugin name and message text) to keep diagnostics stable across refactors.
</issue_to_address>

### Comment 6
<location> `docs/execplans/3-1-1-weaver-plugins-crate.md:14` </location>
<code_context>
+
+## Purpose / big picture
+
+After this change an operator (or AI agent) can register specialist external
+tools as plugins and execute them via the `weaverd` daemon. Actuator plugins
+produce unified diffs that are validated by the existing Double-Lock safety
</code_context>

<issue_to_address>
**nitpick (typo):** Consider adding a comma after "change" for smoother grammar.

For example: "After this change, an operator (or AI agent) can register specialist external tools as plugins..." reads more smoothly.

```suggestion
After this change, an operator (or AI agent) can register specialist external
```
</issue_to_address>

### Comment 7
<location> `crates/weaverd/src/dispatch/act/refactor/mod.rs:108` </location>
<code_context>
+    extra: Vec<String>,
+}
+
+fn parse_refactor_args(arguments: &[String]) -> Result<RefactorArgs, DispatchError> {
+    let mut provider = None;
+    let mut refactoring = None;
</code_context>

<issue_to_address>
**issue (complexity):** Consider inlining the small parsing and path helpers into `parse_refactor_args` and `handle` so all argument and file validation lives in one clear control flow.

You can simplify the argument parsing and path handling without changing behavior by inlining the small helpers and doing all validation in one place.

### 1. Collapse `parse_flag_value` + `validate_required_args` into `parse_refactor_args`

Right now the logic is split across three functions for a single, simple parse. You can keep `RefactorArgs` and the external API the same, but make the control flow easier to follow by doing everything in `parse_refactor_args`:

```rust
fn parse_refactor_args(arguments: &[String]) -> Result<RefactorArgs, DispatchError> {
    let mut provider: Option<String> = None;
    let mut refactoring: Option<String> = None;
    let mut file: Option<String> = None;
    let mut extra = Vec::new();

    let mut iter = arguments.iter().peekable();
    while let Some(arg) = iter.next() {
        match arg.as_str() {
            "--provider" => {
                provider = Some(
                    iter.next()
                        .cloned()
                        .ok_or_else(|| DispatchError::invalid_arguments(
                            "--provider requires a value",
                        ))?,
                );
            }
            "--refactoring" => {
                refactoring = Some(
                    iter.next()
                        .cloned()
                        .ok_or_else(|| DispatchError::invalid_arguments(
                            "--refactoring requires a value",
                        ))?,
                );
            }
            "--file" => {
                file = Some(
                    iter.next()
                        .cloned()
                        .ok_or_else(|| DispatchError::invalid_arguments(
                            "--file requires a value",
                        ))?,
                );
            }
            other => extra.push(other.to_owned()),
        }
    }

    Ok(RefactorArgs {
        provider: provider.ok_or_else(|| {
            DispatchError::invalid_arguments("act refactor requires --provider <plugin-name>")
        })?,
        refactoring: refactoring.ok_or_else(|| {
            DispatchError::invalid_arguments("act refactor requires --refactoring <operation>")
        })?,
        file: file.ok_or_else(|| {
            DispatchError::invalid_arguments("act refactor requires --file <path>")
        })?,
        extra,
    })
}
```

This keeps all the current behavior (same error messages, same `RefactorArgs` shape, same `extra` collection) but removes two indirection layers and makes it clear at a glance what flags are supported and how they’re validated.

### 2. Inline `resolve_file` to keep validation in one place

For this one handler, you can keep all the “validate/resolve file” logic in `handle`, which makes it easier to see the complete flow from arguments to file content:

```rust
// inside handle(...)
let file_path = {
    let path = std::path::Path::new(&args.file);
    if path.is_absolute() {
        return Err(DispatchError::invalid_arguments(
            "absolute file paths are not allowed; use a path relative to the workspace root",
        ));
    }
    let resolved = workspace_root.join(path);
    if !resolved.starts_with(workspace_root) {
        return Err(DispatchError::invalid_arguments("path traversal is not allowed"));
    }
    resolved
};
```

This preserves all checks from `resolve_file` but removes another small helper, so all input validation (flags + file path) is visible in a single function.
</issue_to_address>

### Comment 8
<location> `docs/execplans/3-1-1-weaver-plugins-crate.md:1` </location>
<code_context>
+# Implement the `weaver-plugins` crate with secure broker-plugin IPC
+
+This ExecPlan is a living document. The sections Constraints, Tolerances,
</code_context>

<issue_to_address>
**suggestion (review_instructions):** Expand the IPC acronym on first use in the heading to comply with the acronym-definition requirement.

Consider changing this heading (or adding a short note immediately after it) to introduce the term in full, for example: "Implement the `weaver-plugins` crate with secure broker-plugin inter-process communication (IPC)" so that IPC is defined on first use.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 9
<location> `docs/execplans/3-1-1-weaver-plugins-crate.md:33` </location>
<code_context>
+
+## Context and orientation
+
+Weaver is a client-daemon tool that orchestrates code analysis and
+modification for AI agents. The CLI (`weaver-cli`) connects to the daemon
+(`weaverd`) over Unix sockets using a JSONL protocol. The daemon routes
</code_context>

<issue_to_address>
**suggestion (review_instructions):** Introduce the CLI acronym in full the first time it appears.

On the following line, where CLI is first used — "The CLI (`weaver-cli`) connects to the daemon" — expand it on first mention, for example "The command-line interface (CLI) (`weaver-cli`) connects to the daemon".

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 10
<location> `docs/execplans/3-1-1-weaver-plugins-crate.md:35` </location>
<code_context>
+
+Weaver is a client-daemon tool that orchestrates code analysis and
+modification for AI agents. The CLI (`weaver-cli`) connects to the daemon
+(`weaverd`) over Unix sockets using a JSONL protocol. The daemon routes
+commands through three domains: `observe` (queries), `act` (mutations), and
+`verify` (checks). All `act` commands pass through a Double-Lock safety
</code_context>

<issue_to_address>
**suggestion (review_instructions):** JSONL is used here without being expanded on first use.

Please expand JSONL the first time it appears, for example: "using a JSON Lines (JSONL) protocol" so that the acronym is defined inline.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 11
<location> `docs/execplans/3-1-1-weaver-plugins-crate.md:226` </location>
<code_context>
+
+## Outcomes & retrospective
+
+All acceptance criteria met. The `weaver-plugins` crate compiles, passes
+lint, and has comprehensive unit and BDD tests (56 unit tests + 5 BDD
+scenarios). The `act refactor` command is wired into `weaverd` dispatch and
</code_context>

<issue_to_address>
**suggestion (review_instructions):** BDD is mentioned here without an earlier expansion in the document.

In this paragraph, where "unit and BDD tests" is used, expand BDD on its first occurrence, for example: "unit and behaviour-driven development (BDD) tests".

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaver-plugins/src/error/tests.rs](https://github.com/leynos/weaver/pull/58/files/fa185d025034fbe9dadf028856438c0875906eff#diff-35445690e0f02c7280ccfaf6dbc4615561d720121a40d7e51aefd361d162adfe)

Comment on lines +126 to +140

~~~rust
fn invalid_output_includes_name_and_detail() {
    let error = PluginError::InvalidOutput {
        name: "noisy".into(),
        message: "plugin produced no output on stdout".into(),
    };
    let message = error.to_string();
    assert!(
        message.contains("noisy"),
        "expected plugin name in message: {message}"
    );
    assert!(
        message.contains("no output on stdout"),
        "expected detail in message: {message}"
    );
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=58&biomarker=Code+Duplication&filename=crates%2Fweaver-plugins%2Fsrc%2Ferror%2Ftests.rs&method=)
The module contains 2 functions with similar structure: invalid_output_includes_name_and_detail,spawn_failed_error_message_includes_details

File: crates/weaver-plugins/src/runner/mod.rs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    #[must_use]
+    pub const fn registry(&self) -> &PluginRegistry {
+        &self.registry
+    }
+
+    /// Returns a mutable reference to the plugin registry.
+    #[must_use]
+    pub const fn registry_mut(&mut self) -> &mut PluginRegistry {
+        &mut self.registry
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-plugins/src/runner/mod.rs:118]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/weaver-plugins/src/tests/mod.rs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+impl PluginExecutor for StubExecutor {
+    fn execute(
+        &self,
+        _manifest: &PluginManifest,
+        _request: &PluginRequest,
+    ) -> Result<PluginResponse, PluginError> {
+        Ok(PluginResponse::success(PluginOutput::Empty))
+    }
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-plugins/src/tests/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/weaverd/src/dispatch/act/refactor/mod.rs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        match arg.as_str() {
+            "--provider" => provider = Some(parse_flag_value(&mut iter, "--provider")?),
+            "--refactoring" => refactoring = Some(parse_flag_value(&mut iter, "--refactoring")?),
+            "--file" => file = Some(parse_flag_value(&mut iter, "--file")?),
+            other => extra.push(other.to_owned()),
+        }
+    }
+
+    validate_required_args(provider, refactoring, file, extra)
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaverd/src/dispatch/act/refactor/mod.rs:144]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/weaver-plugins/src/process.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-plugins/src/process.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/weaver-plugins/src/runner/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-plugins/src/runner/mod.rs:118]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/weaver-plugins/src/protocol/mod.rs [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    /// Creates a failed response with diagnostics.
+    #[must_use]
+    pub const fn failure(diagnostics: Vec<PluginDiagnostic>) -> Self {
+        Self {
+            success: false,
+            output: PluginOutput::Empty,
+            diagnostics,
+        }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-plugins/src/protocol/mod.rs:147]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/weaverd/src/dispatch/act/refactor/mod.rs [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+                parts[0].to_owned(),
+                serde_json::Value::String(parts[1].to_owned()),
+            );
+        }
+    }
+
+    let _plugin_request = PluginRequest::with_arguments(
+        &args.refactoring,
+        vec![FilePayload::new(file_path.clone(), file_content)],
+        plugin_args,
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaverd/src/dispatch/act/refactor/mod.rs:83]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaver-plugins/src/tests/behaviour.rs](https://github.com/leynos/weaver/pull/58/files/80d129b95622fa7bcabf52afcf96422bb8e5bc28#diff-5c1f6b40d7ef8b6186f3a6841153b24189cbde3e0f17803c02112b317450d2ad)

Comment on file

~~~rust
//! Behaviour-driven tests for plugin execution.
~~~

❌ New issue: [String Heavy Function Arguments](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=58&biomarker=String+Heavy+Function+Arguments&filename=crates%2Fweaver-plugins%2Fsrc%2Ftests%2Fbehaviour.rs&method=)
In this module, 39.4% of all arguments to its 18 functions are strings. The threshold for string arguments is 39.0%

File: crates/weaver-plugins/src/process.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-plugins/src/process.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
